Clarify how to restore a database backup in cluster (#3035)
neo4j/docs-operations: modules/ROOT/pages/clustering/databases.adoc

9 rows available after 3 ms, consumed after another 1 ms
----

[[restore-database-using-uri-approach]]
=== Restore a database from a URI

> Before restoring a database, ensure you have its backup and drop the database on all servers.

Author: ```suggestion
Before restoring a database from a backup, ensure you have its backup and drop the database on all servers.
```

> Before restoring a database, you have to take its backup and drop the database on all servers.

Author: ```suggestion
Before restoring a database, ensure you have its backup and drop the database on all servers.
```

If you have not dropped the database before restoring it, use `CREATE OR REPLACE DATABASE` instead of the `CREATE DATABASE` command.

The seed can be a full backup, a differential backup, or a dump from an existing database.
When using `CloudSeedProvider`, the URI can point also to a folder which contains a backup chain.
The sources of seeds are called _seed providers_. +
For detailed information about seed providers in Neo4j, see xref:database-administration/standard-databases/seed-from-uri.adoc[].

Remember that the dump file created with the xref:backup-restore/offline-backup.adoc#offline-backup-command[`neo4j-admin database dump`] command contains only the database data.
The associated users and roles metadata are not included and must be recreated manually after restore.

When restoring a database from a backup seed in a cluster, you have to define the database topology.
You have the opportunity to create a topology that will differ from an original one.

> The new allocations can be put on any available server in the cluster.

Author: Available or ENABLED server?

Reviewer: Servers have to be `ENABLED` before they can host user databases; the one exception is the system database which always exists everywhere.

As a final step, you can xref:backup-restore/restore-backup.adoc#_restore_users_and_roles_metadata[restore users and roles metadata].
This step is optional.

Verify that the restored database is online on the desired number of servers, in the desired roles by running the `SHOW DATABASES` command.

Starting with 2026.04, you can also restore a database from a server URI using the `server://` scheme.
See xref:database-administration/standard-databases/seed-from-uri.adoc#server-seed-provider[Create a database from a URI -> `ServerSeedProvider`] for details.

[[restore-database-using-recreate-procedure]]
=== Restore a database using the `dbms.recreateDatabase()` procedure

For detailed information on the recreate procedure, its options and limitations, see xref:database-administration/standard-databases/recreate-database.adoc[].

You need xref:authentication-authorization/dbms-administration/dbms-database-management-privileges.adoc[the `CREATE DATABASE` and `DROP DATABASE` privileges] to run the recreate procedure.

> If you provide a URI to a backup or a dump, the stores on all servers will be replaced by the backup or the dump at the given URI.

Author: Do we have to drop the existing database before calling the recreate procedure?

Reviewer: No you do not.

Possible values are `backup` and `database` (default).
* `--compress` -- label:new[Introduced in 2026.04] (optional) produces compressed backup artifacts.
See xref:backup-restore/copy-database.adoc#copy-database-command-options[Copy a database store] for more information.

`--source-location` and `--source-format` cannot be used together with `--from-path-data` and/or `--from-path-txn` due to the conflict in trying to supply data from two different locations at once.

. Copy the data from a backup of your standard database `foo` into the database `foo-restored`.
+
[source, shell]
----
neo4j-admin database copy foo foo-restored --source-location=s3://bucket/folder/foo.backup --source-format=backup --copy-schema --target-location=s3://bucket/folder --target-format=backup
----
+
Copying a database does not automatically create it.
Therefore, it will not be visible if you do `SHOW DATABASES` at this point.

. Run the following command to create the database `foo-restored` by seeding it from your backup folder in the AWS S3 bucket:
+
[.tabbed-example]
=====
[role=include-with-Cypher-25 label--new-2025.06]
======

[source, cypher]
----

> CREATE DATABASE `foo-restored`

Author: ```suggestion
CREATE DATABASE foo-restored
```

TOPOLOGY [desired number of primaries] PRIMARIES [desired number of secondaries] SECONDARIES
OPTIONS {seedUri: “s3://bucket/folder/”};
----

======
[role=include-with-Cypher-5]
======

[source, cypher]
----
CREATE DATABASE foo-restored
TOPOLOGY [desired number of primaries] PRIMARIES [desired number of secondaries] SECONDARIES
OPTIONS {existingData: 'use', seedUri: “s3://bucket/folder/”};
----
======
=====
+
Number of primaries and secondaries must respect the number of available servers and their mode constraints.

> . Wait until the last committed transaction ID is the same for all allocations of the database and then verify that the new database is online by running the following command:

Author: ```suggestion
. Verify that the new database is online by running the following command:
```

+
[source, cypher]
----
SHOW DATABASES;
----
. After the new database is online, you can drop the old database `foo` if it is no longer needed, or put it into read-write mode if you want to keep it for other purposes.
+
[.tabbed-example]
=====
[.include-with-drop-database]
======

Run the following command to drop the old database `foo`:

[source, cypher]
----
DROP DATABASE foo;
----
======
[.include-with-change-access-mode]
======

Change the access mode of the old database `foo` to read-write:

[source, cypher]
----
ALTER DATABASE foo SET ACCESS READ WRITE;
----
======
=====

> 

Author: What happens to users/roles metadata? Or is it untouched during the copy process?

Reviewer: `neo4j-admin database copy` is maintained by kernel so @EmelieBondesson can give an more informed answer. But my understanding is no, copying a database doesn't copy over the users/roles metadata to the new database.

Reviewer 2: From my understanding we don't copy over the user/roles metadata, we create a new database so they would have to be recreated for that new one if they want the same there as the old database

If you want to ensure a certain allocation across the cluster, you can specify the desired topology with the `TOPOLOGY` clause in the `CREATE DATABASE` command.
See <<create-database, `CREATE DATABASE`>> for more information.
+

> [source, cypher5, role=noplay]

Author: ```suggestion
[source, cypher, role=noplay]
```

To restore a database using backup as a seed and keeping the original database topology, run the command:

[source, shell]
----
CALL dbms.recreateDatabase("foo", {seedURI: "s3://myBucket/myBackup.backup",});
----

If you need to change the original database topology, run the command:

[source, shell]
----
CALL dbms.recreateDatabase("foo",  {seedURI: "s3://myBucket/myBackup.backup", primaries: 3, secondaries: 0});
----

Number of newly defined `primaries` and `secondaries` must respect number of `ENABLED` servers in a cluster and their mode constraints.

Starting with 2026.04, you can also use server URI to recreate a database from a backup/dump located on a server in the cluster's file system:

[source, shell]
----
CALL dbms.recreateDatabase("foo", {seedURI: "server://server-id/myBackup.backup"});
----

If a database is restored using `dbms.recreateDatabase()`, all previously granted user and role privileges on that database are preserved.

As a final step, verify that the restored database is online on the desired number of servers, in the desired roles by running the `SHOW DATABASES` command.

[role=label--new-2026.03]
[[restore-database-using-copy-command]]

> === Restore a database backup using the `neo4j-admin database copy` command

Reviewer 2: This looks copied from the other page for copy so we should remember to apply the changes made there here as well

----

Number of newly defined `primaries` and `secondaries` must respect number of `ENABLED` servers in a cluster and their mode constraints.

Starting with 2026.04, you can also use server URI to recreate a database from a backup/dump located on a server in the cluster's file system:

[source, shell]
----
CALL dbms.recreateDatabase("foo", {seedURI: "server://server-id/myBackup.backup"});
----

If a database is restored using `dbms.recreateDatabase()`, all previously granted user and role privileges on that database are preserved.

As a final step, verify that the restored database is online on the desired number of servers, in the desired roles by running the `SHOW DATABASES` command.

[role=label--new-2026.03]
[[restore-database-using-copy-command]]
=== Restore a database from a backup using the `neo4j-admin database copy` command

See xref:backup-restore/copy-database.adoc#neo4j-admin-copy-command-limitations[Command limitations] before using this option.

The following example assumes:

* You have a backup (created with the `neo4j-admin database backup` command) of your existing database `foo` in an AWS S3 bucket.

> * Or you use the copy command with the new `--target-format=backup --target-location=S3 bucket`.

Reviewer 2: Same here
